### PR TITLE
HPCC-19169 Generate Simplified Expression and check generated expression

### PR DESCRIPTION
**Pull request by shamser**
…expression

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by shamser** [comment]
@ghalliday Please can you review

File: ecl/eclcc/eclcc.cpp [hunk 1]
```diff
@@ -2635,6 +2640,9 @@ int EclCC::parseCommandLineOptions(int argc, const char* argv[])
         else if (iter.matchFlag(optTraceCache, "--tracecache"))
         {
         }
+        else if (iter.matchFlag(optCheckSimpleDef, "--checksimpledef"))
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply ecl/eclcc/eclcc.cpp]: Following on from the offsite discussion this should probably be prefixed with debug or internal (not sure if we agreed a convention) to indicate it it isn't an option we want users to set.
e.g., --internalverifysimplified

File: ecl/hql/hqlgram2.cpp [hunk 2]
```diff
@@ -12234,13 +12233,73 @@ void parseAttribute(IHqlScope * scope, IFileContents * contents, HqlLookupContex
...
+    }
     bool canCache = parsed && (parsed->getOperator() != no_forwardscope);
     bool isMacro = parsed && parsed->isMacro();
 
     attrCtx.noteEndAttribute(true, canCache, isMacro, simplified);
+    if (simplified && (simplified != parsed) && ctx.syntaxChecking())
+        scope->defineSymbol(LINK(simplified));
+}
+
+IHqlExpression * parseDefinition(const char * ecl, IIdAtom * name, MultiErrorReceiver &errors)
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: new functions need a comment to describe what they do - e.g., why does this differ from parseAttribute

File: ecl/hql/hqlgram2.cpp [hunk 3]
```diff
@@ -12234,13 +12233,73 @@ void parseAttribute(IHqlScope * scope, IFileContents * contents, HqlLookupContex
...
+    try
+    {
+        HqlGram parser(scope, scope, contents, attrCtx, NULL, false, true);
+        parser.setExpectedAttribute(name);
+        parser.setAssociateWarnings(true);
+        parser.getLexer()->set_yyLineNo(1);
+        parser.getLexer()->set_yyColumn(1);
+        ::Release(parser.yyParse(false, false));
+    }
+    catch (...)
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: try block can be deleted...

File: ecl/hql/hqlthql.cpp [hunk 4]
```diff
@@ -3447,6 +3479,19 @@ StringBuffer &regenerateECL(IHqlExpression * expr, StringBuffer &s)
 }
 
 
+StringBuffer &regenerateDefinition(IHqlExpression * expr, StringBuffer &s)
+{
+    if (!isEclAlias(expr))
+        return s;
```
Comment (ghalliday) [inline ecl/hql/hqlthql.cpp]: throwUnexpected() would be more appropriate.

**Review by ghalliday — COMMENTED** [review]
Generally looks good a few minor comments.

**Comment by ghalliday** [comment]
@shamser

File: ecl/eclcc/eclcc.cpp [hunk 5]
```diff
@@ -1166,7 +1167,7 @@ void EclCC::processSingleQuery(EclCompileInstance & instance,
 
         optionHash = rtlHash64Data(sizeof(optLegacyImport), &optLegacyImport, optionHash);
         optionHash = rtlHash64Data(sizeof(optLegacyWhen), &optLegacyWhen, optionHash);
-
+        optionHash = rtlHash64Data(sizeof(optCheckSimpleDef), &optCheckSimpleDef, optionHash);
```
[the remark this answers is not in the document]
Comment (shamser) [reply ecl/eclcc/eclcc.cpp]: No, it doesn't change how the query is parsed.  However, the flag does affect what is produced in the contents of the cache file. The reason I added this flag to the hash is that if I were to switch on the 'CheckSimpleDef' flag to get the definition dumped to the cached file, it wouldn't overwrite the existing cache file with a new one that include the IR dump of the simplified expression.

**Review by ghalliday — COMMENTED** [review]
A few comments.

**Comment by ghalliday** [comment]
@shamser a couple of comments.  Please squash before a final review.

File: ecl/hql/hqlgram2.cpp [hunk 6]
```diff
@@ -12176,18 +12176,33 @@ bool parseForwardModuleMember(HqlGramCtx & _parent, IHqlScope *scope, IHqlExpres
...
     attrCtx.noteBeginAttribute(scope, contents, name);
 
-    try
+    // Use simplified definition if syntax checking
+    if (!ctx.regenerateCache() && ctx.syntaxChecking() && ctx.hasCacheLocation() && !ctx.disableCacheUse())
     {
-        //The attribute will be added to the current scope as a side-effect of parsing the attribute.
-        const char * moduleName = scope->queryFullName();
+        HqlParseContext & parseContext = ctx.queryParseContext();
+        Owned<IEclCachedDefinition> cached = parseContext.cache->getDefinition(str(name));
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: This doesn't look right - I would expect it to be the fullname

**Comment by ghalliday** [comment]
@shamser a few failures:

catchexpr.ecl + others
can be fixed by changing to:

        EclIR::getIRText(t1, 0, queryLocationIndependent(simplifiedDefinition));
        EclIR::getIRText(t2, 0, queryLocationIndependent(parsed));

To remove location annotations.
I also added:
            DBGLOG("simple: %s\n parsed: %s", t1.str(), t2.str());
before reporting the error.

import7.ecl:
reportError(1:1) : Name of exported symbol 'm1___1' does not match the expected name 'm1'

And private regression suite:
addressclean1: compile6 and many other examples.
  It looks like the 3rd argument (default parameters) has not been cloned correctly.

It also showed that global dependencies that are gathered if this option is enabled are no longer correct - so we need to check that they are not being overwritten, and think through what happens if they are requested (disable this option??).

**Comment by ghalliday** [comment]
I think I need to rerun my baseline tests.  I also had errors like:

 error C2309: ATMOST(0) doesn't make any sense

Which will need to be moved later as semantic checks.

**Comment by ghalliday** [comment]
A selection of other issues (each of which occur multiple times):
addressclean1b: warning C2029: Constant group/sort clauses make no sense
alima1: warning C1036: Use CHOOSEN(dataset, ALL) to remove implicit choosen.  CHOOSEN(dataset, 0) now returns no records.
alphaslow1: fails unknown id
pig10: error C2167: Unknown identifier "action"

**Comment by shamser** [comment]
@ghalliday Please can you review?

**Comment by ghalliday** [comment]
@shamser Those changes look reasonable.  There are still a large number of problems with the full regression suite.  It might be best to work through some of them together on Wednesday/Thursday.

Running with  -x -syntax -x --nogpg -x --internalverifysimplified

**Comment by shamser** [comment]
@ghalliday Regression issues resolved.  Please can you review.

File: ecl/hql/hqlcache.cpp [hunk 7]
```diff
@@ -369,16 +374,49 @@ static IHqlExpression * createSimplifiedBodyDefinition(IHqlExpression * expr)
...
-  //MORE: check if the default arguments are simple enough (or can be simplfied)
-        return replaceChild(expr->queryBody(), 0, newBody);
+
+        IHqlExpression * params = expr->queryChild(1);
+        HqlExprArray funcArgs;
+        HqlExprArray dummyAttrs;
+        ForEachChild(i, params)
+        {
+            IHqlExpression * param = params->queryChild(i)->queryBody();
+            OwnedHqlExpr newType = createSimplifiedDefinitionType(getFullyUnqualifiedType(param->queryType()));
```
Comment (ghalliday) [inline ecl/hql/hqlcache.cpp]: I'm not sure why this is here.  It means that only integer parameters will be allowed.

File: ecl/hql/hqlcache.cpp [hunk 8]
```diff
@@ -369,16 +374,49 @@ static IHqlExpression * createSimplifiedBodyDefinition(IHqlExpression * expr)
...
+        OwnedHqlExpr formals = createSortList(funcArgs);
+
+        IHqlExpression * origDefaults = expr->queryChild(2);
+        OwnedHqlExpr newDefaults;
+        if (origDefaults && origDefaults->numChildren())
+        {
+            HqlExprArray newDefaultsArray;
+            ForEachChild(idx, origDefaults)
+            {
+                if (origDefaults->queryChild(idx)->getOperator() == no_omitted)
```
Comment (ghalliday) [inline ecl/hql/hqlcache.cpp]: origDefaults->queryChild(idx) should be assigned to a local

File: ecl/hql/hqlcache.cpp [hunk 9]
```diff
@@ -342,7 +342,7 @@ void convertSelectsToPath(StringBuffer & filename, const char * eclPath)
 
 //---------------------------------------------------------------------------------------------------------------------
 
-static IHqlExpression * createSimplifiedDefinition(ITypeInfo * type)
+static IHqlExpression * createSimplifiedDefinitionType(ITypeInfo * type, bool implicitConstantType=false)
```
Comment (ghalliday) [inline ecl/hql/hqlcache.cpp]: not sure I like the name change - it implies it returns a type.  Possibly createSimplifiedDefinitionFromType()

File: ecl/hql/hqlcache.cpp [hunk 10]
```diff
@@ -369,16 +374,49 @@ static IHqlExpression * createSimplifiedBodyDefinition(IHqlExpression * expr)
...
+        IHqlExpression * params = expr->queryChild(1);
+        HqlExprArray funcArgs;
+        HqlExprArray dummyAttrs;
+        ForEachChild(i, params)
+        {
+            IHqlExpression * param = params->queryChild(i)->queryBody();
+            OwnedHqlExpr newType = createSimplifiedDefinitionType(getFullyUnqualifiedType(param->queryType()));
+            if (!newType)
+                return nullptr;
+            funcArgs.append(*createParameter(param->queryId(),(unsigned)expr->querySequenceExtra(), LINK(newType->queryType()), dummyAttrs));
```
Comment (ghalliday) [inline ecl/hql/hqlcache.cpp]: minor: newType->getType() would be more concise.

File: ecl/hql/hqlcache.cpp [hunk 11]
```diff
@@ -354,13 +354,18 @@ static IHqlExpression * createSimplifiedDefinition(ITypeInfo * type)
         //Possible, but the default testing code doesn't work
         return nullptr;
     case type_int:
+        if (implicitConstantType)
```
Comment (ghalliday) [inline ecl/hql/hqlcache.cpp]: This should have a comment to indicate why this is being done.

File: ecl/hql/hqlcache.cpp [hunk 12]
```diff
@@ -369,16 +374,49 @@ static IHqlExpression * createSimplifiedBodyDefinition(IHqlExpression * expr)
...
+                }
+            }
+            newDefaults.set(createSortList(newDefaultsArray));
+        }
+        return createFunctionDefinition(expr->queryId(), newBody.getClear(), formals.getClear(), newDefaults.getClear(), nullptr);
     }
     if (expr->getOperator()==no_typedef || expr->getOperator()==no_enum)
         return nullptr;
-    ITypeInfo * type = expr->queryType();
+    ITypeInfo * type = getFullyUnqualifiedType(expr->queryType());
```
Comment (ghalliday) [inline ecl/hql/hqlcache.cpp]: This will leak.

**Comment by ghalliday** [comment]
@shamser various comments

**Comment by shamser** [comment]
@ghalliday Please can you review.

File: ecl/hql/hqlexpr.cpp [hunk 13]
```diff
@@ -1070,77 +1070,80 @@ bool HqlParseContext::checkEndMeta()
     return wasGathering;
 }
 
-void HqlParseContext::finishMeta(bool isSeparateFile, bool success, bool generateMeta, bool canCache, bool isMacro, IHqlExpression * simplifiedDefinition)
+void HqlParseContext::createCache(IHqlExpression * simplifiedDefinition, bool isMacro)
 {
-    if (metaStack.empty())  // paranoid - could only happen on an internal error
-        return;
+    assertex(simplifiedDefinition);
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: I'm not sure if this is correct (e.g., for macros).  I have added HPCC-19623 to ensure it is revisited later.

**Comment by ghalliday** [comment]
@shamser a few final comments.  I think we should also disable generating of cache entries if we are compiling from an archive.  (But still allow them to be verified.) I can't think of a reason it would be useful, and I can foresee it causing problems.

File: ecl/hql/hqlcache.cpp [hunk 14]
```diff
@@ -341,8 +341,16 @@ void convertSelectsToPath(StringBuffer & filename, const char * eclPath)
...
+// type - type to be simplified
+// implicitConstantType - when false, generate the IR with same type as 'type'
+//                      - otherwise generate the IR with the implicit type of a constant of type 'type'
+//                        For example, for type UINT8 generate constant INT8 as the numeric '0' will normally
+//                        be parsed as this type.  This is used where the parser uses the implicit type when
+//                        explicit type cannot be determined. They are used, for example, in function
+//                        parameter's default value:  func(UINT4 arg=0).  The parser would parse the default value
+//                        for the parameter as an INT8.  The simplified expression constant should be INT8
+//                        rather than UINT4 to ensure that the simplified expression is consistent.
+static IHqlExpression * createSimplifiedDefinitionFromType(ITypeInfo * type, bool implicitConstantType=false)
```
[the remark this answers is not in the document]
Comment (shamser) [reply ecl/hql/hqlcache.cpp]: Could there be implicit constants of other types?

File: ecl/hql/hqlcache.cpp [hunk 15]
```diff
@@ -341,8 +341,16 @@ void convertSelectsToPath(StringBuffer & filename, const char * eclPath)
 }
 
 //---------------------------------------------------------------------------------------------------------------------
-
-static IHqlExpression * createSimplifiedDefinitionType(ITypeInfo * type, bool implicitConstantType=false)
+// type - type to be simplified
+// implicitConstantType - when false, generate the IR with same type as 'type'
```
[the remark this answers is not in the document]
Comment (shamser) [reply ecl/hql/hqlcache.cpp]: I thought it was shortcut for expression tree.  I got it from dump_ir().

**Comment by shamser** [comment]
@ghalliday The location of the meta has been reverted to the original location

**Pull request merged**